### DISCUSSION TITLE
Allow individual `start` hook for @astrojs/node

### DISCUSSION THREAD

**Discussion by pReya** [withastro/roadmap#967] [Proposal]
### Body

# Summary

Not sure if this is "big enough" for its own RFC. But this is where GitHub directed me. It might just be a minor PR.

It would be very helpful, if the `@astrojs/node` would allow to pass a function as a `start` parameter, which would then be executed when the server is started. 

Specifying a `start` hook is already a [feature of the Adapter API](https://docs.astro.build/en/reference/adapter-reference/#start) – it would only be necessary to forward `start` via the Adapter configuration.

# Background & Motivation

This would be useful to setup the proper server environment before the Astro application is started. E.g. setting up a cron job.

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- Allow individual user scripts to be executed when the Astro node adapter is started

# Example

```

// https://astro.build/config
export default defineConfig({
  // Enable React to support React JSX components.
  integrations: [
    react()
  ],
  output: "server",
  adapter: node({
    mode: "standalone",
    start: () => {console.log("This will be executed before the server is started")}
  }),
});

**Comment by ematipico** [comment]
What kind of payload would you expect from this hook?

I would also argue that passing a function like this isn't a safe approach, because this is information that must be available after a build, and a function can't because it's not serialisable. Usually, we accept a path to an entry point, load it, and execute it.

**Reply by pReya** [reply]
I'd suggest it to be identical to the [`start` function of the Adapter API](https://docs.astro.build/en/reference/adapter-reference/#start).

**Comment by delucis** [comment]
Isn’t this already possible with a simple server script something like:

```js
// ./run-server.mjs
console.log("This will be executed before the server is started");
await import("./dist/server/entry.mjs");
```

```sh
node ./run-server.mjs
```

**Reply by pReya** [reply]
Sure, there's many workarounds to this. But they're typically easier to brake and not as comfortable as passing a function through the config file.

Would it make sense to add this solution to the docs, somewhere?

**Reply by delucis** [reply]
I’m not a huge user of the Node adapter, so probably not best placed to say if my suggested workaround is the right one or not to document.

---

As @ematipico mentioned above, a function directly inline in the Astro config probably wouldn’t work, because the Astro config is not bundled and run in the server output. If this were ever supported, it would probably look more like an option to tell Astro to use a specific file:

```js
node({ startEntrypoint: './start.js' })
```

**Reply by ematipico** [reply]
> Sure, there's many workarounds to this. But they're typically easier to brake and not as comfortable as passing a function through the config file.
> 
> Would it make sense to add this solution to the docs, somewhere?

If this `start` method doesn't pass any information that belongs to Astro, where users can do something based on that "information," I can't see enough value in adding it, considering that you can do it yourself with a simple script.

**Reply by pReya** [reply]
> > Sure, there's many workarounds to this. But they're typically easier to brake and not as comfortable as passing a function through the config file.
> > Would it make sense to add this solution to the docs, somewhere?
> 
> If this `start` method doesn't pass any information that belongs to Astro, where users can do something based on that "information," I can't see enough value in adding it, considering that you can do it yourself with a simple script.

It's a common feature of many existing full-stack frameworks, to run a function once, when starting the application (e.g. Laravel, Rails, Django). I can understand your objections/the problems with the hook inside the config. But mentioning the "workaround" of writing a simple script for it could be a useful addition to the docs, don't you think?

**Reply by ematipico** [reply]
That's definitely worth adding to the docs!!

**Reply by Madd0g** [reply]
sorry, I might have missed something - did you mean that creating the wrapper file that imports `entry.mjs` is what's worth adding to the docs?

I'm wondering what other workarounds exist and are recommended?

**Reply by dvelasquez** [reply]
The wrapper seems like an interesting idea.

As of today I have a warmup cache function that load some huge file files in memory using cacheable. The only place we found so far was executing this function in the middleware.ts file but the problem with this approach is that it runs once when the first request is received.

Using the wrapper you suggest could do this as well, although I imagina that is not a script that will be compiled by the astro build, so I would have to compile it manually (if it is typescript) and copy it to the dist folder.

What would be nice is if the build process of astro took care of this.

Another option I imagine would be to implement a custom server using express or fastify and in that script add my warmup function, but I would like to avoid implementing custom servers, in the past with other frameworks (NextJS) has been a nightmare to keep the compatibility when the framework is updated so I would prefer to stick to the official adapter.

**Reply by edvinasjurele** [reply]
Same here! As discussed in duplicate discussion about the same topic https://github.com/withastro/roadmap/discussions/1005

We are using it for initializing some SDKs:
- Sentry (though this one we have under custom Fastify-driven `server.mjs` level, but could be moved into app)
- Redis SDK - for establishing connections and making it ready to work
- AB experiment tool SDK
- other SDKs

We had been using the `middlware/index.ts` file before because it was running during initialization and this worked on 4.x, but was removed in v5.0.5 with https://github.com/withastro/astro/pull/12707/files as it was bug, though for us it was a feature :D

Given that, since the middleware runs with the first GET request (either page or API endpoint) and not on server init, we get Redis initialized later too.

I really think having some hook, as mentioned previously:
```javascript
node({ startEntrypoint: './start.js' }) // mayber .ts ?
```
might be a way to go.

**Reply by mistyharsh** [reply]
:wave: Checking after months!
Has there been any official or community-wide accepted convention for this?

**Reply by smnbbrv** [reply]
@edvinasjurele 

> we had been using the middlware/index.ts file before because it was running during initialization and this worked on 4.x

I did the same thing :) until came to a solution

**Comment by smnbbrv** [comment]
Hi guys, as a one who was suffering from this problem, I invested some time into a solution.

https://github.com/smnbbrv/astroscope/tree/main/packages/boot

Please let me know what you think.

**Comment by dvelasquez** [comment]
While working on monitoring and profiling I stumble to a very old way to achieve this.

In node you can do this:

`node --require ./src/this-loads-first.js ./src/index.js`

This will load `./src/this-loads-first.js` before `./src/index.js`

In a ESM world, you can achieve this by using `import` instead of `require`:

`node --import ./src/this-loads-first.js ./src/index.js`

You can even add multiple `--imports`, they go on that exact same order.

What if you want to use typescript?

You can use the following:

`node --experimental-strip-types --import ./src/this-loads-first.ts ./src/index.ts`.

Full example for a node astro app:

`node --experimental-strip-types --import ./src/preload.ts ./dist/server/entryPoint.mjs`

If you want to preload something in your dev script (`npm run dev`) you can use the NODE_OPTIONS env, like this:

`"dev": "NODE_OPTIONS=\"--experimental-strip-types --import ./src/preload.ts\" astro dev`.

Hope it works and happy coding!
